stft: move towards always returning complex

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #72882

For `stft` this makes all cases where `return_complex` default to
`False` into an error and adds a warning when `return_complex=False`
is passed explicitly.

For `istft` this raises an error if the input is not a complex tensor.

peterbell10 committed:
stft: move towards always returning complex

For `stft` this makes all cases where `return_complex` default to
`False` into an error and adds a warning when `return_complex=False`
is passed explicitly.

For `istft` this raises an error if the input is not a complex tensor.

[ghstack-poisoned]

peterbell10 committed:
Update on "stft: move towards always returning complex"


For `stft` this makes all cases where `return_complex` default to
`False` into an error and adds a warning when `return_complex=False`
is passed explicitly.

For `istft` this raises an error if the input is not a complex tensor.

[ghstack-poisoned]

peterbell10 committed:
Update on "stft: move towards always returning complex"


For `stft` this makes all cases where `return_complex` default to
`False` into an error and adds a warning when `return_complex=False`
is passed explicitly.

For `istft` this raises an error if the input is not a complex tensor.

[ghstack-poisoned]

mruberry commented on torch/csrc/jit/operator_upgraders/upgraders_entry.cpp:
```diff
@@ -15,6 +15,27 @@ namespace torch {
+                     length=length, return_complex=return_complex)
+)SCRIPT"},
+    {"stft_0_10", R"SCRIPT(
+def stft_0_10(self: Tensor, n_fft: int, hop_length: Optional[int]=None, win_length: Optional[int]=None,
```
@tugsbayasgalan should know what to do

peterbell10 replied:
#71993 removes the `istft` python wrapper but is blocked by a JIT bug (#71994). `stft` depends on `torch.nn.functional.pad` which is implemented in python.

tugsbayasgalan replied:
We currently iterate over all nodes in the graph and replace the aten operators with upgraders when necessary. I am wondering if we can be little smarter in that pass that would dig deeper when we encounter `prim::CallFunction`. I am not sure how feasible that is @eellison @gmagogsfm. Alternatively, fixing the bug @anjali411 mentioned to unblock the  moving to C++ might be the way to go. What do you guys think?

peterbell10 committed:
Fix istft merge conflict on "stft: move towards always returning complex"


For `stft` this makes all cases where `return_complex` default to
`False` into an error and adds a warning when `return_complex=False`
is passed explicitly.

For `istft` this raises an error if the input is not a complex tensor.

[ghstack-poisoned]

peterbell10 commented:
Sure thing. The actual code shouldn't need changed, but I'm still waiting on #73434 which is blocked by an FC period that ends this week,

anjali411 commented:
@peterbell10 we should merge this PR right?

peterbell10 commented:
I should have said, the deprecations have been re-submitted as #86724 and #86628 which have been merged.
